File: 0704-binary-search/0704-binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1
        
        # '<=' because we have to check the midpoint
        while low <= high:
            midpoint = (low + high)//2 # '//' represents integer division
            if nums[midpoint] == target:
                return midpoint
            elif target < nums[midpoint]:
                high = midpoint - 1
            else: # target > nums[midpoint]
                low = midpoint + 1
        
        return -1
        

File: 0704-binary-search/0704-binary-search.c
// assume a sorted array
int search(int* nums, int numsSize, int target) {
    int *i = nums;
    int *j = nums + (numsSize-1);
    while (i <= j) {
        int *midpoint = i + ((int)(j-i) / 2);
        if (*midpoint == target) {
            return (int)(midpoint-nums);
        } else if (*midpoint > target) {
            j = midpoint - 1;
        } else { // *midpoint < target
            i = midpoint + 1;
        }
    }
    return -1;
}
